### ISSUE TITLE
Add an event handler for plugins when the editor receive a new state

### ISSUE THREAD

**Issue by SamyPesse** [ianstormtaylor/slate#218]
I'm making progress on the schema validation ( [slate-schema](https://github.com/SamyPesse/slate-schema)).

But something is missing in the API to build a "validation" plugin: When the editor receive a new state modified by another part of the application (for example a Redux reducer), there is no event handler that can be called to process this new state before rendering.

Then a wrong state will be display before the user type something or move the selection, and the `onChange` is called to validate the state.

I propose to add a `onRender` or `onNewState` or `onReceiveState` event handler that will be chained like `onChange`.

----

PS-1: I know that it's possible to do:

```js
const myValidationPlugin = MyPlugin(...);

render() {
     return <Slate.Editor
          onChange={...}
          state={myValidationPlugin.onChange(this.state.state}
          plugins={[myValidationPlugin]}
    />;
}
```

But I think Slate can provide a better API.

----

PS-2: I understand that in theory it should not needed since the developer should not build transforms that can break its state, but when building a large rich text edition, preventing all the cases during transform can easily become a huge pain.

**Comment by ianstormtaylor** [comment]
Interesting, I definitely understand the use case, but I also agree that it feels hackish from the theory standpoint. Do you have a few use cases in mind where this has come up already? Would be helpful to make sure I'm thinking about it correctly.

Also, I wonder if a solution might be to call it `onBeforeChange` and have it fire before any `onChange` handler, but including when passing a new state, so that "normalization" would only need to be implemented in that handler alone.

**Comment by SamyPesse** [comment]
Closed by #219
